c-dependencies/js-compute-runtime/js-compute-builtins.cpp

```diff
@@ -3322,6 +3322,341 @@ JSObject *create_rs_proxy(JSContext *cx, HandleObject input_readable) {
+const size_t BUFFER_SIZE = 16384;
+
+bool is_instance(JSObject *obj);
+
+JSObject *transform(JSObject *self) {
+  MOZ_ASSERT(is_instance(self));
+  return &JS::GetReservedSlot(self, Slots::Transform).toObject();
+}
+
+Format format(JSObject *self) {
+  MOZ_ASSERT(is_instance(self));
+  return (Format)JS::GetReservedSlot(self, Slots::Format).toInt32();
+}
+
+z_stream *state(JSObject *self) {
+  MOZ_ASSERT(is_instance(self));
+  void *ptr = JS::GetReservedSlot(self, Slots::State).toPrivate();
+  MOZ_ASSERT(ptr);
+  return (z_stream *)ptr;
+}
+
+uint8_t *output_buffer(JSObject *self) {
+  MOZ_ASSERT(is_instance(self));
+  void *ptr = JS::GetReservedSlot(self, Slots::Buffer).toPrivate();
+  MOZ_ASSERT(ptr);
+  return (uint8_t *)ptr;
+}
+
+const unsigned ctor_length = 1;
+bool check_receiver(JSContext *cx, HandleValue receiver, const char *method_name);
+
+// Steps 1-5 of the transform algorithm, and 1-5 of the flush algorithm.
+bool inflate_chunk(JSContext *cx, HandleObject self, HandleValue chunk, bool finished) {
+  z_stream *zstream = state(self);
+
+  if (!finished) {
+    // 1.  If _chunk_ is not a `BufferSource` type, then throw a `TypeError`.
+    // Step 2 of transform:
+    size_t length;
+    uint8_t *data = value_to_buffer(cx, chunk, "DecompressionStream transform: chunks", &length);
+    if (!data) {
+      return false;
+    }
+
+    if (length == 0) {
+      return true;
+    }
+
+    // 2.  Let _buffer_ be the result of decompressing _chunk_ with _ds_'s format
+    // and context. This just sets up step 2. The actual decompression happen in
+    // the `do` loop below.
+    zstream->avail_in = length;
+
+    // `data` is a live view into `chunk`. That's ok here because it'll be fully
+    // used in the `do` loop below before any content can execute again and
+    // could potentially invalidate the pointer to `data`.
+    zstream->next_in = data;
+  } else {
+    // Step 1 of flush:
+    // 1.  Let _buffer_ be the result of decompressing an empty input with _ds_'s
+    // format and
+    //     context, with the finish flag.
+
+    // Step 2 of flush:
+    // 2.  If the end of the compressed input has not been reached, then throw a TypeError.
+    if (zstream->avail_in != 0) {
+      JS_ReportErrorNumberUTF8(cx, GetErrorMessage, nullptr, JSMSG_DECOMPRESSING_ERROR);
+      return false;
+    }
+    // This just sets up step 3. The actual decompression happens in the `do` loop
+    // below.
+    zstream->avail_in = 0;
+    zstream->next_in = nullptr;
+  }
+
+  RootedObject controller(cx, TransformStream::controller(transform(self)));
+
+  // Steps 3-5 of transform are identical to steps 3-5 of flush, so numbers
+  // below refer to the former for those. Also, the compression happens in
+  // potentially smaller chunks in the `do` loop below, so the three steps are
+  // reordered and somewhat intertwined with each other.
+
+  uint8_t *buffer = output_buffer(self);
```

Comment:
Are private values not managed by the spidermonkey gc, and as a result don't need to be rooted?

Reply:
I assume they are not managed, my assumption is based upon the fact that our CompressionStream implementation does the same thing - https://github.com/fastly/js-compute-runtime/blob/890b51709bedadd857d3e7e332d73061a7363fb1/c-dependencies/js-compute-runtime/js-compute-builtins.cpp#L3762